jax/experimental/checkify.py

```diff
@@ -0,0 +1,269 @@
+
+from dataclasses import dataclass
+from functools import partial
+import itertools as it
+from typing import Union, Optional, Callable, Dict
+
+import numpy as np
+
+import jax.numpy as jnp
+
+from jax import core
+from jax import linear_util as lu
+from jax.api_util import flatten_fun
+from jax.interpreters import partial_eval as pe
+from jax.tree_util import tree_flatten, tree_unflatten, register_pytree_node
+from jax._src import source_info_util, traceback_util
+from jax._src.lax import lax
+from jax._src.lax import slicing
+from jax._src.lax import control_flow
+from jax._src.util import as_hashable_function, unzip2
+
+source_info_util.register_exclusion(__file__)
+traceback_util.register_exclusion(__file__)
+
+
+## Utils
+
+def popattr(obj, attrname):
+  val = getattr(obj, attrname)
+  delattr(obj, attrname)
+  return val
+
+def setnewattr(obj, name, val):
+  sentinel = object()
+  assert getattr(obj, name, sentinel) is sentinel
+  setattr(obj, name, val)
+
+## Error value data type and functional assert.
+
+@dataclass(frozen=True)
+class Error:
+  err: Union[bool, core.Tracer]
+  code: Union[int, core.Tracer]
+  msgs: Dict[int, str]
+
+  def get(self) -> Optional[str]:
+    assert np.shape(self.err) == np.shape(self.code)
+    if np.size(self.err) == 1:
+      if self.err:
+        return self.msgs[int(self.code)]
+    else:
+      return '\n'.join(f'at mapped index {", ".join(map(str, idx))}: '  # type: ignore
+                       f'{self.msgs[int(self.code[idx])]}'              # type: ignore
+                       for idx, e in np.ndenumerate(self.err) if e) or None
+    return None
+
+register_pytree_node(Error,
+                     lambda e: ((e.err, e.code), tuple(sorted(e.msgs.items()))),
+                     lambda msgs, data: Error(*data, dict(msgs)))  # type: ignore
+
+init_error = Error(False, 0, {})
+next_code = it.count(1).__next__  # globally unique ids, could be uuid4
+
+
+Bool = Union[bool, core.Tracer]
+Int = Union[int, core.Tracer]
+
+def assert_func(error: Error, pred: Bool, msg: str) -> Error:
+  code = next_code()
+  out_err = error.err | jnp.logical_not(pred)
+  out_code = lax.select(error.err, error.code, code)
+  return Error(out_err, out_code, {code: msg, **error.msgs})
+
+## Checkify transformation for plumbing functional error values.
+
+class ErrorTrace(core.Trace):
+  pure = lift = sublift = lambda self, val: ErrorTracer(self, val)
+
+  def process_primitive(self, primitive, tracers, params):
+    in_vals = [t.val for t in tracers]
+    rule = error_checks.get(primitive)
+    if rule:
+      out, self.main.error = rule(self.main.error, *in_vals, **params)
+    else:
+      out = primitive.bind(*in_vals, **params)
+    if primitive.multiple_results:
+      return [ErrorTracer(self, x) for x in out]
+    else:
+      return ErrorTracer(self, out)
+
+  def process_call(self, primitive, f, tracers, params):
+    in_vals = [t.val for t in tracers]
+    e = popattr(self.main, 'error')
```

Comment:
Why do we need to pop the attribute here (and later set it again?)

Reply:
We want the error value (i.e. the boolean and code) to be threaded through the call in the sense that it should be passed as an argument to the call application (in the jaxpr) and returned as an output.

When not dealing with higher-order primitives like this, we just keep the error value stuck on the MainTrace instance, as global data. (A cleaner version of this is `global_data` in Autodidax, an explicit attribute on MainTrace, but it's the same idea.)

So at this line in the code we want to get the error value from the MainTrace instance so that we can pass it in as an argument to the called function. Then just on the "other side" of the call, we take that passed-in argument and glue it back on the MainTrace instance. In short, we take off the error value, pass it through the call as an explicit argument, then immediately glue it back on.

In that first step, we don't actually have to pop/delete the error attribute from the MainTrace instance, but I thought it was cleaner that way.

WDYT?

Reply:
yeah, I guess my confusion was about why we pop it instead of getting it, I guess to make sure that the error is not available as an attribute on the Maintrace within the call primitive? Just to be safe?

Reply:
Yeah exactly, just to be safe. (Sorry for mostly answering the wrong question!)